### PR TITLE
Sync system revision

### PR DESCRIPTION
**Pull request by DavidCramer**
Revises the sync system to be thread controlled, and introduces WP CLI sync commands.

### CODE REVIEW

File: php/sync/class-sync-queue.php [hunk 1]
```diff
@@ -440,20 +652,50 @@ protected function schedule_resume() {
...
 	}
 
+	/**
+	 * Get the state of the thread.
+	 *
+	 * @param string $thread Thread name to check.
+	 *
+	 * @return int  0 = disabled, 1 = ended, 2 = active, 3 = stalled/ready to start.
+	 */
+	public function get_thread_state( $thread ) {
```
Comment (pereirinha) [inline php/sync/class-sync-queue.php:662]: @DavidCramer 

This is a great one!

**Review by pereirinha — COMMENTED** [review]
@DavidCramer 

I left you a few comments.
Do you mind to check them and tell me what you think?

Thank you.

File: php/class-cli.php [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+	 * @param array|bool $include Include already found items.
+	 *
+	 * @return array
+	 */
+	private function get_all_unsynced( $include = false ) {
+
+		$params = array(
+			'post_type'      => 'attachment',
+			'post_status'    => 'inherit',
+			'fields'         => 'ids',
```
[the remark this answers is not in the document]
Comment (DavidCramer) [reply php/class-cli.php]: @pereirinha - Agreed on the cache items. However, we use the found_rows for the progress bar.
I've refactored the whole CLI to make it easier to use and to reduce load and insanity.

File: php/sync/class-sync-queue.php [hunk 3]
```diff
@@ -47,241 +63,228 @@ class Sync_Queue {
...
 	 * Upload_Queue constructor.
 	 *
 	 * @param \Cloudinary\Plugin $plugin The plugin.
 	 */
 	public function __construct( \Cloudinary\Plugin $plugin ) {
-		$this->plugin = $plugin;
-
-		$this->cron_frequency    = apply_filters( 'cloudinary_cron_frequency', 10 * MINUTE_IN_SECONDS );
+		$this->plugin            = $plugin;
+		$this->cron_frequency    = apply_filters( 'cloudinary_cron_frequency', 10 );
```
[the remark this answers is not in the document]
Comment (DavidCramer) [reply php/sync/class-sync-queue.php:93]: @pereirinha - oops, I set it for testing, forgot to set it back. Good catch!

File: php/sync/class-sync-queue.php [hunk 4]
```diff
@@ -294,144 +297,353 @@ public function validate_queue() {
...
-		return $queue;
+		$query = new \WP_Query( $args );
+		if ( ! $query->have_posts() ) {
+			// translators: variable is page number.
+			$action_message = sprintf( __( 'No posts', 'cloudinary' ), $args['paged'] );
+			do_action( '_cloudinary_queue_action', $action_message );
+
+			return;
+		}
+		$ids = array_merge( $query->get_posts(), $ids );
```
[the remark this answers is not in the document]
Comment (DavidCramer) [reply php/sync/class-sync-queue.php:323]: @pereirinha - good catch. Left over from when it was looping.

File: php/class-cli.php [hunk 5]
```diff
@@ -0,0 +1,148 @@
...
+		}
+		$total = count( $assets );
+		\WP_CLI::log( '' );
+		\WP_CLI::log( \WP_CLI::colorize( '%gStarting Sync:%n' ) );
+		$bar = \WP_CLI\Utils\make_progress_bar( 'Syncing ' . $total . ' assets', $total, 10 );
+		foreach ( $assets as $index => $asset ) {
+			$file = get_attached_file( $asset );
+			$bar->tick( 0, 'Syncing: ' . basename( $file ) . ' (' . ( $index + 1 ) . ' of ' . $total . ')' );
+			$this->plugin->get_component( 'sync' )->managers['push']->process_assets( $asset, $bar );
+			$bar->tick();
```
[the remark this answers is not in the document]
Comment (DavidCramer) [reply php/class-cli.php:184]: @pereirinha - Refactored a little.

File: php/sync/class-sync-queue.php [hunk 6]
```diff
@@ -294,144 +297,353 @@ public function validate_queue() {
...
+	protected function get_thread_queue_details( $thread ) {
+
+		$args = array(
+			'post_type'      => 'attachment',
+			'post_mime_type' => array( 'image', 'video' ),
+			'post_status'    => 'inherit',
+			'posts_per_page' => 1,
+			'fields'         => 'ids',
+			'cache_results'  => false,
+			'meta_query'     => array( // phpcs:ignore
```
[the remark this answers is not in the document]
Comment (DavidCramer) [reply php/sync/class-sync-queue.php:540]: Updated

**Comment by DavidCramer** [comment]
@pereirinha - Ready to look at this again.

File: php/class-cli.php [hunk 7]
```diff
@@ -7,142 +7,238 @@
...
+			// Setup process.
+			$total   = $query->found_posts;
+			$process = "process_{$process}";
+			do {
+				$posts = $query->get_posts();
+				$this->{$process}( $posts, $total );
+
+				// Free up memory.
+				if ( method_exists( $this, 'stop_the_insanity' ) ) {
+					$this->stop_the_insanity();
```
Comment (pereirinha) [inline php/class-cli.php:151]: @DavidCramer 

We need to implement this method as it's not part of the WP_CLI.

**Review by pereirinha — COMMENTED** [review]
@DavidCramer 

I left you a comment

**Pull request merged**